source/common/network/lc_trie.h

```diff
@@ -0,0 +1,567 @@
+     * @param prefix supplies the prefix to ignore when building the sub-trie.
+     * @param first supplies the index into ip_prefixes_ for this sub-trie.
+     * @param n suppplies the number of entries for the sub-trie.
+     * @param position supplies the root for this sub-trie.
+     * @param next_free_index supplies the next available index in the trie_.
+     */
+    void buildRecursive(uint32_t prefix, uint32_t first, uint32_t n, uint32_t position,
+                        uint32_t& next_free_index) {
+      // Setting a leaf, the branch and skip are 0.
+      if (n == 1) {
+        // There is no way to predictably determine the number of trie nodes required to build a
+        // LC-Trie. If while building the trie the position that is being set exceeds the maximum
+        // number of supported entries, throw an Envoy Exception instead of letting an out_of_range
+        // exception be thrown.
+        // TODO(ccaraman): Add a test case.
+        if (position > MAXIMUM_TRIE_NODES) {
+          throw EnvoyException(fmt::format("The number of internal nodes required for the LC-Trie "
+                                           "exceeded the maximum number of "
+                                           "supported nodes. Number of internal nodes required: "
+                                           "'{0}'. Maximum number of supported nodes: '{1}'.",
+                                           position, MAXIMUM_TRIE_NODES));
+        }
+        trie_[position].address_ = first;
+        return;
+      }
+
+      ComputePair output = computeBranchAndSkip(prefix, first, n);
+
+      uint32_t address = next_free_index;
+      trie_[position].branch_ = output.branch_;
+      // The skip value is the number of bits between the newly calculated prefix(output.prefix_)
+      // and the previous prefix(prefix).
+      trie_[position].skip_ = output.prefix_ - prefix;
+      trie_[position].address_ = address;
+
+      // The next available free index to populate in the trie_ is at next_free_index + 2^(branching
+      // factor).
+      next_free_index += 1 << output.branch_;
+
+      uint32_t new_position = first;
+
+      // Build the subtrees.
+      for (uint32_t bit_pattern = 0; bit_pattern < static_cast<uint32_t>(1 << output.branch_);
+           ++bit_pattern) {
+
+        // count is the number of entries in the base vector that have the same bit pattern as the
+        // ip_prefixes_[new_position].
+        int count = 0;
+        while (new_position + count < first + n &&
+               static_cast<uint32_t>(extractBits<IpType, address_size>(
+                   output.prefix_, output.branch_, ip_prefixes_[new_position + count].ip_)) ==
+                   bit_pattern) {
+          ++count;
+        }
+
+        // This logic was taken from
+        // https://github.com/beevek/libkrb/blob/24a224d3ea840e2e7d2926e17d8849aefecc1101/krb/lc_trie.hpp#L396.
```

Comment:
I think this section lacks in explanation compared to the rest of the code, and it is a pretty crucial set of operations. I'll leave it up to you if you want to spell it out more. At the very least to explain what each case is.

Reply:
Added more comments